### PR TITLE
Fix: Controls do not implement a value of 0 [ED-7935]

### LINKED ISSUE #19410: 🐞 Bug report: 3.7 release when Widget Space in columns is set to "0" it reverts to 20px default on live view
### Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] The issue still exists against the latest stable version of Elementor.

### Description

When widget space in columns is set to "0" it reverts to 20px default on live view but not in editor. All the spacing on our sites that had "0" set is now 20px. 

If I change it to a differen't number besides 0 it works, like 1, or 5, but when set to zero it shows 20px spacing after publishing but shows correctly within the editor. 

### Steps to reproduce

Add elements to a column, set column widget space to "0", save and view live page and you'll see it reverts to 20px.

### Isolating the problem

- [X] This bug happens with only Elementor plugin active (and Elementor Pro).
- [X] This bug happens with a Blank WordPress theme active ([Hello theme](https://wordpress.org/themes/hello-elementor/)).
- [X] I can reproduce this bug consistently using the steps above.

### System Info

					
== Server Environment ==
	Operating System: Linux
	Software: Flywheel/5.1.0
	MySQL version: (Google) v5.7.37
	PHP Version: 7.4.30
	PHP Memory Limit: 128M
	PHP Max Input Vars: 10000
	PHP Max Post Size: 300M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: There are some writing permissions issues with the following directories/files:
		 - WordPress root directory
	Elementor Library: Connected

== WordPress Environment ==
	Version: 6.0.1
	Site URL: https://splash.mysites.io
	Home URL: https://splash.mysites.io
	WP Multisite: No
	Max Upload Size: 300 MB
	Memory limit: 256M
	Max Memory limit: 256M
	Permalink Structure: /%postname%/
	Language: en-US
	Timezone: 0
	Admin Email: ben@apolloartistry.com
	Debug Mode: Inactive

== Theme ==
	Name: Apollox
	Version: 3.2.0
	Author: Apollo Artistry
	Child Theme: Yes
	Parent Theme Name: Astra
	Parent Theme Version: 3.9.1
	Parent Theme Author: Brainstorm Force

== User ==
	Role: administrator
	WP Profile lang: en_US
	User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36

== Active Plugins ==
	Astra Pro
		Version: 3.9.1
		Author: Brainstorm Force

	Elementor
		Version: 3.7.0
		Author: Elementor.com

	Elementor Pro
		Version: 3.7.3
		Author: Elementor.com

	JetElements For Elementor
		Version: 2.6.7
		Author: Crocoblock

	JetTricks
		Version: 1.4.2
		Author: Crocoblock

	Ultimate Addons for Elementor
		Version: 1.36.8
		Author: Brainstorm Force

== Elementor Experiments ==
	Optimized DOM Output: Inactive by default
	Improved Asset Loading: Inactive by default
	Improved CSS Loading: Inactive by default
	Inline Font Icons: Inactive by default
	Accessibility Improvements: Inactive by default
	Additional Custom Breakpoints: Inactive by default
	Import Export Template Kit: Active by default
	Hide native WordPress widgets from search results: Active by default
	admin_menu_rearrangement: Inactive by default
	Flexbox Container: Inactive by default
	Default to New Theme Builder: Active by default
	Landing Pages: Active by default
	Color Sampler: Active by default
	Favorite Widgets: Active by default
	Admin Top Bar: Inactive
	Page Transitions: Active by default
	Notes: Inactive
	Form Submissions: Active by default
	Scroll Snap: Inactive

== Log ==
	
Log: showing 20 of 422022-07-14 18:08:11 [info] Elementor/Upgrades - _on_each_version Start  
2022-07-14 18:08:11 [info] Elementor/Upgrades - _on_each_version Finished 
2022-07-14 18:08:11 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.6.4',
  'to' => '3.6.7',
)]
2022-07-14 18:08:12 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.6.5',
  'to' => '3.7.2',
)]
2022-07-14 18:08:23 [info] elementor-pro::elementor_pro_updater Started 
2022-07-14 18:08:23 [info] Elementor Pro/Upgrades - _on_each_version Start  
2022-07-14 18:08:23 [info] Elementor Pro/Upgrades - _on_each_version Finished 
2022-07-14 18:08:23 [info] Elementor Pro/Upgrades - _v_3_7_2_woocommerce_rename_related_to_related_products Start  
2022-07-14 18:08:23 [info] Elementor Pro/Upgrades - _v_3_7_2_woocommerce_rename_related_to_related_products Finished 
2022-07-14 18:08:23 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.6.5',
  'to' => '3.7.2',
)]
2022-08-09 22:52:41 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.6.7',
  'to' => '3.7.0',
)]
2022-08-09 22:52:55 [info] elementor::elementor_updater Started 
2022-08-09 22:52:55 [info] Elementor/Upgrades - _on_each_version Start  
2022-08-09 22:52:55 [info] Elementor/Upgrades - _on_each_version Finished 
2022-08-09 22:52:55 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.6.7',
  'to' => '3.7.0',
)]
2022-08-09 23:08:16 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.7.2',
  'to' => '3.7.3',
)]
2022-08-09 23:10:38 [info] elementor-pro::elementor_pro_updater Started 
2022-08-09 23:10:38 [info] Elementor Pro/Upgrades - _on_each_version Start  
2022-08-09 23:10:38 [info] Elementor Pro/Upgrades - _on_each_version Finished 
2022-08-09 23:10:38 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.7.2',
  'to' => '3.7.3',
)]

PHP: showing 1 of 1PHP: 2022-04-14 18:31:10 [notice X 3][/www/wp-content/plugins/elementor/modules/dev-tools/deprecation.php::301] The $control_id argument is deprecated since version 3.5.0! [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]

== Elementor - Compatibility Tag ==
	
	Elementor Pro: Compatible
	JetElements For Elementor: Incompatible
	Ultimate Addons for Elementor: Incompatible

== Elementor Pro - Compatibility Tag ==
	
	JetElements For Elementor: Incompatible
	Ultimate Addons for Elementor: Compatible

### PR DESCRIPTION
**Pull request by udidol**
## PR Type
What kind of change does this PR introduce?

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

* Fix: Controls do not implement a value of 0 (Fixes #19410 #19391 #19393 #19386)

### CODE REVIEW

File: core/files/css/base.php [hunk 1]
```diff
@@ -14,6 +14,7 @@
 use Elementor\Plugin;
 use Elementor\Stylesheet;
 use Elementor\Icons_Manager;
+use Elementor\Utils;
```
Comment (ManorHazaz) [inline core/files/css/base.php:17]: Remove

File: tests/phpunit/elementor/core/files/css/test-base.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+namespace Elementor\Tests\Phpunit\Elementor\Core\Files\Css;
+
+use ElementorEditorTesting\Elementor_Test_Base;
+
+/**
+ * Test the CSS Base class
+ */
+class Test_Base extends Elementor_Test_Base {
+
+	public function test_parse_property_placeholder() {
```
Comment (ManorHazaz) [inline tests/phpunit/elementor/core/files/css/test-base.php:11]: This test should test the typical situation when you pass any standard value and not an edge case, so change the name or change the test :)

Comment (ManorHazaz) [reply tests/phpunit/elementor/core/files/css/test-base.php:11]: Also add tests for all the edge cases: "any_type_of_empty", "0" and 0 (as you already did).

File: tests/phpunit/elementor/core/files/css/test-base.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		// Create mock data.
+		$control = [
+			'type' => 'number',
+			'default' => '20',
+		];
+
+		$controls_stack = [
+			'number' => $control,
+		];
+
```
Comment (ManorHazaz) [inline tests/phpunit/elementor/core/files/css/test-base.php:25]:
```suggestion

// Act.
```

File: tests/phpunit/elementor/core/files/css/test-base.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+		//array $control, $value, array $controls_stack, $value_callback, $placeholder, $parser_control_name = null
+		$control_value = $css_base_class->parse_property_placeholder(
+			$control,
+			0,
+			$controls_stack,
+			function() {},
+			''
+		);
+
+		// Assert. Check that when the control value is 0, `parse_property_placeholder()` returns 0.
```
Comment (ManorHazaz) [inline tests/phpunit/elementor/core/files/css/test-base.php:35]: The explanation is redundant, the code is simple.

**Review by ghost — DISMISSED** [review]

File: tests/phpunit/elementor/core/files/css/test-base.php (hunk 2 above)
Comment (udidol) [reply tests/phpunit/elementor/core/files/css/test-base.php]: Done

File: core/files/css/base.php (hunk 1 above)
Comment (udidol) [reply core/files/css/base.php]: Done

File: tests/phpunit/elementor/core/files/css/test-base.php (hunk 3 above)
Comment (udidol) [reply tests/phpunit/elementor/core/files/css/test-base.php:25]: Done

File: tests/phpunit/elementor/core/files/css/test-base.php (hunk 4 above)
Comment (udidol) [reply tests/phpunit/elementor/core/files/css/test-base.php]: Done

File: core/files/css/base.php [hunk 5]
```diff
@@ -442,8 +441,8 @@ public function parse_property_placeholder( array $control, $value, array $contr
 			$value = call_user_func( $value_callback, $control );
 		}
 
-		// If the control value is empty, check for global default.
-		if ( ! $value ) {
+		// If the control value is empty, check for global default. `0` (integer, string) are falsy but are valid values.
+		if ( empty( $value ) && '0' !== $value && 0 !== $value ) {
```
Comment (StyleShit) [inline core/files/css/base.php:445]: Maybe use `Utils::is_empty()`?

File: tests/phpunit/elementor/core/files/css/test-base.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	 */
+	private static $mock_control;
+	/**
+	 * @var array[]
+	 */
+	private static $mock_controls_stack;
+
+	public static function setUpBeforeClass() {
+		// The CSS Base class is abstract, so it can't be instantiated. The inheriting Post class is used instead.
+		self::$css_base_class = new \Elementor\Core\Files\CSS\Post( 0 );
```
Comment (StyleShit) [inline tests/phpunit/elementor/core/files/css/test-base.php:26]: It's a better practice to have a fresh instance for each test instead of reusing the same one

File: tests/phpunit/elementor/core/files/css/test-base.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+			'type' => 'number',
+			'default' => 20,
+		];
+
+		self::$mock_controls_stack = [
+			'number' => self::$mock_control,
+		];
+	}
+
+	public function test_parse_property_placeholder__value_0() {
```
Comment (StyleShit) [inline tests/phpunit/elementor/core/files/css/test-base.php:38]:
```suggestion
	public function test_parse_property_placeholder__value_0_as_integer() {
```

File: tests/phpunit/elementor/core/files/css/test-base.php [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		self::$css_base_class = new \Elementor\Core\Files\CSS\Post( 0 );
+
+		self::$mock_control = [
+			'type' => 'number',
+			'default' => 20,
+		];
+
+		self::$mock_controls_stack = [
+			'number' => self::$mock_control,
+		];
```
Comment (StyleShit) [inline tests/phpunit/elementor/core/files/css/test-base.php:35]: same as before, it's better for future-devs readability to have all of the "arrange" part in the test itself instead of having to look for it. In this case it's not a big deal, but when this file will have tons of tests it will be

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: core/files/css/base.php (hunk 5 above)
Comment (udidol) [reply core/files/css/base.php:445]: Not good. It handles arrays in a way that is not compatible for the use here

File: tests/phpunit/elementor/core/files/css/test-base.php (hunk 8 above)
Comment (udidol) [reply tests/phpunit/elementor/core/files/css/test-base.php]: I changed the method to the non-static `setUp()` instead of `setUpBeforeClass()`.
When more tests will be added to this file, this can be refactored to an "arrange helper" function instead of `setUp()` so it doesn't run for unnecessary tests. But for now I think this is fine.

File: tests/phpunit/elementor/core/files/css/test-base.php (hunk 7 above)
Comment (udidol) [reply tests/phpunit/elementor/core/files/css/test-base.php]: Changed

File: tests/phpunit/elementor/core/files/css/test-base.php (hunk 6 above)
Comment (udidol) [reply tests/phpunit/elementor/core/files/css/test-base.php]: Done

**Review by StyleShit — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: core/files/css/base.php (hunk 5 above)
Comment (matipojo) [reply core/files/css/base.php:445]: Why double checking? `&& '0' !== $value && 0 !== $value`

Comment (udidol) [reply core/files/css/base.php:445]: Because I specifically want to handle these two cases.
Casting to an `int` in this comparison can return true for non-wanted values such as `false`, or even `'false'`.
